master/buildbot/process/build.py

```diff
@@ -530,28 +529,38 @@ def startNextStep(self):
             return self.allStepsDone()
         self.executedSteps.append(s)
         self.currentStep = s
-        d = defer.maybeDeferred(s.startStep, self.conn)
-        d.addBoth(self._flushProperties)
-        d.addCallback(self._stepDone, s)
-        d.addErrback(self.buildException)
-        return None
+
+        # the following function returns a deferred, but we don't wait for it
+        self._start_next_step_impl(s)
+        return defer.succeed(None)
 
     @defer.inlineCallbacks
-    def _flushProperties(self, results):
-        # `results` is just passed on to the next callback
-        yield self.master.data.updates.setBuildProperties(self.buildid, self)
+    def _start_next_step_impl(self, step):
+        try:
+            results = yield step.startStep(self.conn)
+            yield self.master.data.updates.setBuildProperties(self.buildid, self)
 
-        return results
+            self.currentStep = None
+            if self.finished:
+                return  # build was interrupted, don't keep building
 
-    @defer.inlineCallbacks
-    def _stepDone(self, results, step):
-        self.currentStep = None
-        if self.finished:
-            return  # build was interrupted, don't keep building
-        terminate = yield self.stepDone(results, step)  # interpret/merge results
-        if terminate:
-            self.terminate = True
-        yield self.startNextStep()
+            terminate = yield self.stepDone(results, step)  # interpret/merge results
+            if terminate:
+                self.terminate = True
+            yield self.startNextStep()
+
+        except Exception as e:
+            log.msg("{} build got exception when running step {}".format(self, step))
```

Comment:
the coverage report shows this is not covered. We should probably have something to cover it.

Reply:
Agreed, I'll write a test for that.

Reply:
Fixed now.